storage2/src/state/transaction.rs

```diff
@@ -0,0 +1,43 @@
+use std::collections::HashMap;
+
+use anyhow::Result;
+
+use jmt::storage::{TreeReader, TreeWriter};
+
+use super::{State, StateWrite};
+
+/// Represents a transactional set of changes to a `State` fork,
+/// implemented as a RYW cache over a `State`.
+pub(crate) struct Transaction<'a, R: TreeReader + TreeWriter> {
+    // TODO: should higher-level types be used here instead of jmt types?
+    cache: HashMap<jmt::KeyHash, jmt::OwnedValue>,
+    unwritten_changes: Vec<(jmt::KeyHash, jmt::OwnedValue)>,
+    state: &'a State<'a, R>,
+}
+
+impl<'a, R: TreeReader + TreeWriter> Transaction<'a, R> {
+    pub fn new() -> Self {
+        Self {
+            cache: todo!(),
+            unwritten_changes: todo!(),
+            state: todo!(),
+        }
+    }
+
+    pub fn commit(self) -> Result<()> {
+        // Write unwritten_changes to the state
+        // `self` will be consumed afterwards
+        todo!()
+    }
+}
+
+impl<'a, R: TreeReader + TreeWriter> StateWrite for Transaction<'a, R> {
+    fn put(&mut self, key: jmt::KeyHash, value: Option<jmt::OwnedValue>) -> Transaction<'a, R> {
+        todo!()
```

Comment:
As with StateRead, we should be taking string keys, which will have to be owned values (hence String).

We'll also want a similar API, with put, put_proto, and put_raw.

Taking the value as an Option is a bad API for general purpose use like this, because it forces every caller to Option-wrap their data, and the reader of every call site to understand that put(Some) means "actually put" while put(None) means "delete". Instead, we should just have a distinct delete() method.

Reply:
We also shouldn't return the transaction. Since we're modifying it with &mut self, the caller doesn't need a return value, and the Transaction isn't cloneable.